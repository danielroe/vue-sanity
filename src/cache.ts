import type { VueConstructor } from 'vue'
import {
  computed,
  getCurrentInstance,
  isRef,
  onServerPrefetch,
  reactive,
  Ref,
  set,
  unref,
  watch,
} from 'vue-demi'

/**
 * Cached data, status of fetch, timestamp of last fetch, error
 */
type CacheEntry<T> = [T, FetchStatus, number, any, Promise<T>]

const cache = reactive<Record<string, CacheEntry<any>>>({})

export function ensureInstance() {
  const instance = getCurrentInstance()
  if (!instance) throw new Error('You must call this from within a component')
  return instance.proxy as InstanceType<VueConstructor>
}

export function getServerInstance() {
  const instance = getCurrentInstance()

  if (instance?.proxy && (process?.env?.VITE_SSG || instance?.proxy.$isServer))
    return instance.proxy as InstanceType<VueConstructor>
  return false
}

export type FetchStatus =
  | 'initialised'
  | 'loading'
  | 'server loaded'
  | 'client loaded'
  | 'error'

interface SetCacheOptions<T, K> {
  key: string
  value?: T | K
  status?: FetchStatus
  error?: any
  promise?: Promise<T> | null
  time?: number
}

export interface CacheOptions<K> {
  initialValue?: K
  /**
   * Whether to disable SSR fetching. Defaults to false. Now replaced with strategy: 'client'
   * @deprecated
   */
  clientOnly?: boolean
  /**
   * Strategy for fetching. Defaults to 'both'.
   * 'server' will not refetch if the cache has been populated on SSR.
   * 'client' will disable SSR fetching.
   * 'both' will fetch on server and refetch when page is loaded.
   */
  strategy?: 'server' | 'client' | 'both'
  /**
   * Whether to de-duplicate identical fetches. If set to `true`, additional fetches will not run unless made after
   * the previous request errors or succeeds. If set to a number, additional fetches will run, but only after this
   * many milliseconds after the previous fetch began.
   */
  deduplicate?: boolean | number
}

export function useCache<T, K = null>(
  key: string | Ref<string>,
  fetcher: (key: string) => Promise<T>,
  options: CacheOptions<K> = {}
) {
  const {
    initialValue = null,
    deduplicate = false,
    strategy = 'both',
    clientOnly = false,
  } = options

  const enableSSR = !clientOnly && strategy !== 'client'

  function initialiseCache({
    key,
    value,
    error = null,
    status = 'initialised',
    time = new Date().getTime(),
  }: SetCacheOptions<T, K>) {
    set(cache, key, [value, status, time, error])
  }

  const serverInstance = getServerInstance()
  const k = unref(key)

  if (enableSSR && !serverInstance) {
    const prefetchState =
      (window as any).__VSANITY_STATE__ ||
      ((window as any).__NUXT__ && (window as any).__NUXT__.vsanity)
    if (prefetchState && prefetchState[k]) {
      const [value, status, time, error] = prefetchState[k] as CacheEntry<T>

      initialiseCache({
        key: k,
        value,
        status,
        time,
        error,
      })
    }
  }

  function verifyKey(key: string) {
    const emptyCache = !(key in cache)
    if (emptyCache) initialiseCache({ key, value: initialValue as K })
    return emptyCache || cache[key][1] === 'initialised'
  }

  function setCache({
    key,
    value = cache[key]?.[0] || initialValue,
    status = cache[key]?.[1],
    error = null,
    promise = cache[key]?.[4],
  }: SetCacheOptions<T, K>) {
    if (!(key in cache)) initialiseCache({ key, value, status })

    set(cache[key], 0, value)
    set(cache[key], 1, status)
    set(cache[key], 2, new Date().getTime())
    set(cache[key], 3, error)
    set(cache[key], 4, promise)
  }

  function fetch(query = unref(key), force?: boolean) {
    if (
      !force &&
      query &&
      cache[query] &&
      cache[query][1] !== 'error' &&
      (cache[query][0] !== initialValue ||
        cache[query][4] instanceof Promise) &&
      deduplicate &&
      (deduplicate === true ||
        deduplicate < new Date().getTime() - cache[query][2])
    )
      return cache[query][4] instanceof Promise
        ? (cache[query][4] as Promise<T>)
        : (Promise.resolve(cache[query][0]) as Promise<T>)

    const promise = fetcher(query)
    setCache({ key: query, status: 'loading', promise })

    promise
      .then(value =>
        setCache({
          key: query,
          value,
          status: serverInstance ? 'server loaded' : 'client loaded',
          promise: null,
        })
      )
      .catch(error => setCache({ key: query, status: 'error', error }))
    return promise
  }

  if (enableSSR && serverInstance) {
    const ctx = serverInstance.$ssrContext
    if (ctx) {
      if (ctx.nuxt && !ctx.nuxt.vsanity) {
        ctx.nuxt.vsanity = {}
      } else if (!ctx.vsanity) {
        ctx.vsanity = {}
      }
    }

    onServerPrefetch(async () => {
      const k = unref(key)
      try {
        await fetch(k, verifyKey(k))
        // eslint-disable-next-line
      } catch {}
      if (
        ctx &&
        cache[k] &&
        !['loading', 'initialised'].includes(cache[k]?.[1])
      ) {
        if (ctx.nuxt) {
          ctx.nuxt.vsanity[k] = cache[k].slice(0, 3)
        } else {
          ctx.vsanity[k] = cache[k].slice(0, 3)
        }
      }
    })
  }

  const data = computed(() => {
    const k = unref(key)
    if (!k) return initialValue

    verifyKey(k)

    return (cache[k] && cache[k][0]) as T | K
  })

  const status = computed(() => {
    const k = unref(key)
    if (!k) return 'server loaded'

    verifyKey(k)

    return cache[k][1]
  })

  const error = computed(() => {
    const k = unref(key)
    if (!k) return null

    verifyKey(k)

    return cache[k][3]
  })

  if (isRef(key)) {
    watch(
      key,
      key => {
        if (strategy === 'server' && status.value === 'server loaded') return

        fetch(key, verifyKey(key))
      },
      { immediate: true }
    )
  } else if (strategy !== 'server' || status.value !== 'server loaded') {
    fetch(key, verifyKey(key))
  }

  return {
    setCache,
    triggerFetch: fetch,
    fetch,
    data,
    status,
    error,
  }
}
