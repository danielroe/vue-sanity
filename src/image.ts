import type { InjectionKey, Ref } from 'vue'
import { computed, inject } from 'vue'

import type { FitMode, SanityImageDimensions } from '@sanity/image-url/lib/types/types'
import type { ImageUrlBuilder } from '@sanity/image-url/lib/types/builder'

export const imageBuilderSymbol: InjectionKey<ImageUrlBuilder> = Symbol('Sanity image URL builder')

export interface ResolvedSanityImage {
  url: string
  dimensions?: SanityImageDimensions
}

interface ImageOptions {
  dpr?: number
  quality?: number
  fit?: FitMode
}

/**
 *
 * WARNING. Work in progress. API may change.
 */
export function useSanityImage(
  image: Ref<ResolvedSanityImage>,
  options?: Partial<ImageOptions>,
  widths = [300, 600, 1200, 1920],
) {
  const builder = inject(imageBuilderSymbol)

  if (!builder) {
    throw new Error(
      'You must call useSanityClient before using sanity resources in this project.',
    )
  }

  function getImageUrl(
    image: Required<ResolvedSanityImage>,
    width: number,
    { quality = 82, fit = 'min' }: ImageOptions,
  ) {
    return (builder as ImageUrlBuilder)
      .image(image.url)
      .width(Math.round(width))
      .quality(quality)
      .fit(fit)
      .url()
  }

  const result = computed(() => ({
    src: image.value.url,
    ...(image.value.dimensions
      ? {
          srcset: [
            ...widths.map(
              width =>
                `${getImageUrl(
                  image.value as Required<ResolvedSanityImage>,
                  width,
                  options || {},
                )} ${width}w`,
            ),
            `${image.value.url} ${image.value.dimensions.width}w`,
          ].join(', '),
          placeholder: '',
        }
      : {
          srcset: [
            ...widths.map(
              width =>
                `${getImageUrl(
                  image.value as Required<ResolvedSanityImage>,
                  width,
                  options || {},
                )} ${width}w`,
            ),
          ].join(', '),
          placeholder: '',
        }),
  }))

  return result
}
