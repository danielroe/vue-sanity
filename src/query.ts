import {
  computed,
  inject,
  watch,
  InjectionKey,
  Ref,
  isRef,
} from '@vue/composition-api'
import minifier from 'minify-groq'
import type { SanityClient } from '@sanity/client'
import type { QueryBuilder } from 'sanity-typed-queries'

import { useCache, CacheOptions, FetchStatus } from './cache'

export interface Client {
  fetch: (query: string) => Promise<any>
  [key: string]: any
}

export const clientSymbol: InjectionKey<Client> = Symbol('Sanity client')
export const previewClientSymbol: InjectionKey<SanityClient> = Symbol(
  'Sanity client for previews'
)
export const optionsSymbol: InjectionKey<Options> = Symbol(
  'Default query options'
)

type Query = string | (() => string | null | undefined | false)

interface Result<T> {
  /**
   * An automatically synced and updated result of the Sanity query.
   */
  data: Ref<T>
  /**
   * The status of the query. Can be 'server loaded', 'loading', 'client loaded' or 'error'.
   */
  status: Ref<FetchStatus>
  /**
   * An error returned in the course of fetching
   */
  error: any
  /**
   * Get result directly from fetcher (integrates with cache)
   */
  fetch: () => Promise<T>
}

export type Options = Omit<CacheOptions<any>, 'initialValue'> & {
  /**
   * Whether to listen to real-time updates from Sanity. You can also pass an object of options to pass to `client.listen`. Defaults to false.
   */
  listen?: boolean | Record<string, any>
}

/**
 *
 * @param query A string, or a function that retuns a query string. If the return value changes, a new Sanity query will be run and the return value automatically updated.
 */
export function useSanityFetcher<T extends any>(query: Query): Result<T | null>

/**
 *
 * @param query A function that retuns a query string. If the return value changes, a new Sanity query will be run and the return value automatically updated.
 * @param initialValue The value to return before the Sanity client returns an actual result. Defaults to null.
 * @param mapper A function that transforms the result from Sanity, before returning it to your component.
 */
export function useSanityFetcher<T extends any, R extends any = T>(
  query: Query,
  initialValue: R,
  mapper?: (result: any) => T,
  options?: Options
): Result<T | R>

export function useSanityFetcher(
  query: Query,
  initialValue = null,
  mapper = (result: any) => result,
  queryOptions?: Options
) {
  const client = inject(clientSymbol)
  const defaultOptions = inject(optionsSymbol, {})

  const options = {
    ...defaultOptions,
    initialValue,
    ...queryOptions,
  }
  if (!client)
    throw new Error(
      'You must call useSanityClient before using sanity resources in this project.'
    )

  const computedQuery =
    typeof query === 'string'
      ? minifier(query).replace(/\n/g, ' ')
      : computed(() => minifier(query() || '').replace(/\n/g, ' '))

  const { data, status, setCache, error, fetch } = useCache(
    computedQuery,
    query =>
      query ? client.fetch(query).then(mapper) : Promise.resolve(initialValue),
    options
  )

  if (options.listen) {
    const previewClient = inject(previewClientSymbol, client as SanityClient)
    if ('listen' in previewClient) {
      const listenOptions =
        typeof options.listen === 'boolean' ? undefined : options.listen

      const subscribe = (query: string) =>
        previewClient.listen(query, listenOptions).subscribe(
          event =>
            event.result &&
            setCache({
              key: query,
              value: event.result,
            })
        )
      if (isRef(computedQuery)) {
        watch(
          computedQuery,
          query => {
            const subscription = subscribe(query)

            const unwatch = watch(
              computedQuery,
              newQuery => {
                if (newQuery !== query) {
                  subscription.unsubscribe()
                  unwatch()
                }
              },
              { immediate: true }
            )
          },
          { immediate: true }
        )
      } else {
        subscribe(computedQuery)
      }
    }
  }

  return { data, status, error, fetch }
}

export function useSanityQuery<
  Builder extends Pick<
    QueryBuilder<Schema, Mappings, Type, Project, Exclude>,
    'use'
  >,
  Schema,
  Mappings extends Record<string, any>,
  Type,
  Project extends boolean,
  Exclude extends string
>(
  builder: Builder | (() => Builder)
): Result<
  ReturnType<Builder['use']>[1] extends Array<any>
    ? ReturnType<Builder['use']>[1]
    : ReturnType<Builder['use']>[1] | null
>

export function useSanityQuery<
  Builder extends Pick<
    QueryBuilder<Schema, Mappings, Type, Project, Exclude>,
    'use'
  >,
  Schema,
  Mappings extends Record<string, any>,
  Type,
  Project extends boolean,
  Exclude extends string
>(
  builder: Builder | (() => Builder),
  initialValue: null
): Result<ReturnType<Builder['use']>[1] | null>

export function useSanityQuery<
  Builder extends Pick<
    QueryBuilder<Schema, Mappings, Type, Project, Exclude>,
    'use'
  >,
  Schema,
  Mappings extends Record<string, any>,
  Type,
  Project extends boolean,
  Exclude extends string,
  InitialValue
>(
  builder: Builder | (() => Builder),
  initialValue: InitialValue
): Result<ReturnType<Builder['use']>[1] | InitialValue>

export function useSanityQuery<
  Builder extends Pick<
    QueryBuilder<Schema, Mappings, Type, Project, Exclude>,
    'use'
  >,
  Schema,
  Mappings extends Record<string, any>,
  Type,
  Project extends boolean,
  Exclude extends string,
  Mapper extends (result: ReturnType<Builder['use']>[1]) => any
>(
  builder: Builder | (() => Builder),
  initialValue: null,
  mapper: Mapper,
  options?: Options
): Result<ReturnType<Mapper> | null>

export function useSanityQuery<
  Builder extends Pick<
    QueryBuilder<Schema, Mappings, Type, Project, Exclude>,
    'use'
  >,
  Schema,
  Mappings extends Record<string, any>,
  Type,
  Project extends boolean,
  Exclude extends string,
  InitialValue,
  Mapper extends (result: ReturnType<Builder['use']>[1]) => any
>(
  builder: Builder | (() => Builder),
  initialValue: InitialValue,
  mapper: Mapper,
  options?: Options
): Result<ReturnType<Mapper> | InitialValue>

export function useSanityQuery<
  Builder extends Pick<
    QueryBuilder<Schema, Mappings, Type, Project, Exclude>,
    'use'
  >,
  Schema,
  Mappings extends Record<string, any>,
  Type,
  Project extends boolean,
  Exclude extends string
>(
  builder: Builder | (() => Builder),
  initialValue = null,
  mapper = (result: any) => result,
  options?: Options
) {
  const query =
    'use' in builder ? () => builder.use()[0] : () => builder().use()[0]
  const type = 'use' in builder ? builder.use()[1] : builder().use()[1]

  return useSanityFetcher<typeof type>(
    query,
    initialValue || type,
    mapper,
    options
  )
}
