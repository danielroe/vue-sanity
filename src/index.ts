import type { ClientConfig } from '@sanity/client'

import type { CacheOptions, FetchStatus } from './cache'
import type { Client, Options } from './query'
import { createClient } from '@sanity/client'

import imageUrlBuilder from '@sanity/image-url'
import { inject, provide } from 'vue'
import { ensureInstance, useCache } from './cache'
import { imageBuilderSymbol, useSanityImage } from './image'
import { clientSymbol, optionsSymbol, previewClientSymbol, useSanityFetcher, useSanityQuery } from './query'

interface RequiredConfig {
  /**
   * Your project ID. You can find it in your sanity.json.
   */
  projectId: string
  /**
   * You must specify which dataset to use. You can find it in your sanity.json.
   */
  dataset: string
}

/**
 * @param config The configuration for the client.
 * @param supportPreview Whether to create a preview client (that won't use CDN, and supports credentials for viewing drafts). Defaults to false.
 * @param defaultOptions Default options for the fetcher.
 */
export function useSanityClient(
  config: ClientConfig & RequiredConfig,
  supportPreview = false,
  defaultOptions: Options = {},
) {
  ensureInstance()

  const client = createClient(config)
  const imageBuilder = imageUrlBuilder(config)

  provide(clientSymbol, client)
  provide(imageBuilderSymbol, imageBuilder)
  provide(optionsSymbol, defaultOptions)

  if (supportPreview) {
    const previewClient = createClient({
      ...config,
      useCdn: false,
      token: undefined,
      withCredentials: true,
    })
    provide(previewClientSymbol, previewClient)
  }
}

export function useCustomClient(client: Client, defaultOptions: Options = {}) {
  ensureInstance()
  provide(clientSymbol, client)
  provide(optionsSymbol, defaultOptions)
}

export function fetch(query: string) {
  ensureInstance()
  const client = inject(clientSymbol)
  if (!client) {
    throw new Error(
      'You must call useSanityClient before using sanity resources in this project.',
    )
  }
  return client.fetch(query)
}

export { useCache, useSanityFetcher, useSanityImage, useSanityQuery }
export type { CacheOptions, FetchStatus, Options }
