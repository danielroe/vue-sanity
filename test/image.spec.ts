import { ref } from 'vue'
import type { ClientConfig } from '@sanity/client'
import { describe, expect, it, vi } from 'vitest'

import { useSanityClient, useSanityImage } from '../src'
import { runInSetup } from './helpers/mount'

const config = {
  projectId: 'id',
  dataset: 'production',
  useCdn: true,
  apiVersion: '2021-03-25',
} satisfies ClientConfig

;(globalThis.console.error as any) = vi.fn()

describe('image builder', () => {
  const image = {
    url: 'https://cdn.sanity.io/images/siteid/production/imageid-100x100.jpg',
    dimensions: {
      aspectRatio: 3,
      height: 3,
      width: 3,
    },
  }

  it('errors without proper config', async () => {
    let error: unknown
    runInSetup(() => {
      try {
        useSanityImage(ref(image))
      }
      catch (e) {
        error = e
      }
      return {}
    })

    expect(error).toBeDefined()
    expect(console.error).toBeCalled()
  })

  it('errors when run outside of setup', async () => {
    let error: unknown
    try {
      useSanityImage(ref(image))
    }
    catch (e) {
      error = e
    }
    expect(error).toBeDefined()
  })

  it('produces relevant image url', async () => {
    const data = await runInSetup(() => useSanityClient(config, true), () => ({
      result: useSanityImage(ref(image)),
    }))

    expect(data.result.value).toMatchSnapshot()
  })

  it('works without dimensions', async () => {
    const data = await runInSetup(() => useSanityClient(config, true), () => ({
      result: useSanityImage(ref({ ...image, dimensions: undefined })),
    }))

    expect(data.result.value).toMatchSnapshot()
  })
})
