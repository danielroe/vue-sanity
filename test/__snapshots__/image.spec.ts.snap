// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`image builder > produces relevant image url 1`] = `
{
  "placeholder": "",
  "src": "https://cdn.sanity.io/images/siteid/production/imageid-100x100.jpg",
  "srcset": "https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?w=300&q=82&fit=min 300w, https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?w=600&q=82&fit=min 600w, https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?w=1200&q=82&fit=min 1200w, https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?w=1920&q=82&fit=min 1920w, https://cdn.sanity.io/images/siteid/production/imageid-100x100.jpg 3w",
}
`;

exports[`image builder > works without dimensions 1`] = `
{
  "placeholder": "",
  "src": "https://cdn.sanity.io/images/siteid/production/imageid-100x100.jpg",
  "srcset": "https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?w=300&q=82&fit=min 300w, https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?w=600&q=82&fit=min 600w, https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?w=1200&q=82&fit=min 1200w, https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?w=1920&q=82&fit=min 1920w",
}
`;
